content/blog/kube-dns-coredns-perfcomp.md

```diff
@@ -0,0 +1,87 @@
+
+**Authoritative Base Kube-dns and CoreDNS Manifests**:  I have based the configurations of kube-dns and CoreDNS on their [authoritative base templates](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns).  Specifically, tests use the authoritative default configurations for kube-dns (dnsmasq command flags in the Deployment) and CoreDNS (Corefile in the Configmap). A single instance of each DNS server was deployed.
+
+**Kubernetes Cluster**: The tests were performed in a three node kubernetes cluster, using the Calico network plugin.  The Nodes were hosted at bare metal provider [Packet](https://www.packet.net), using a TBD as the master, and TBD for the nodes.
+
+**Kubernetes DNS Perf-test Tool**: I used the [kubernetes dns perf-tests tool](https://github.com/kubernetes/perf-tests/tree/master/dns) modified to [allow tests on CoreDNS](https://github.com/kubernetes/perf-tests/pull/114).  The tool by default runs the DNS server under test on the master node and the client pod on separate node (node 1).  The tests are not performed against the cluster's active cluster DNS, instead they configure, spin up and tear down an instance for each set of tests.
+
+**Upstream Server**: The upstream server was a CoreDNS instance using the `template` plugin, configured to reply to all queries with answer of '1.2.3.4'.  It was run in the cluster, on a separate node from the client and DNS server under test (node 2).  This local server responds much more quickly than a real world upstream server would.
+
+**DNS Server Versions**: CoreDNS 1.1.3, kube-dns 1.14.10
+
+
+## CoreDNS and Kube-dns QPS and Latency
+
+Each set of tests is for 3 types of `A` record queries:
+
+1. **invalid**: A name for a service that does not exist in the cluster, which prompts an `NXDOMAIN` response from the DNS service. e.g. `invalid-service.default.svc.cluster.local`
+2. **service**: A name for a service that exists in the cluster. e.g. `kubernetes.default.svc.cluster.local`
+3. **upstream**: A name outside of the cluster, which is forwarded upstream by the DNS service. e.g. `coredns.io`
+
+
+### Results
+
+TODO
+
+## Cluster-First DNS Policy
+
+Recall that in Kubernetes, queries for out-of-cluster names (e.g. google.com) result in at least 3 futile in-cluster queries, followed by the correct upstream query.
+
+Factoring in the Cluster-First DNS policy into the restricted CPU performance results, we can project performance from the client pod perspective ...
+
+### Out Of Cluster Cache Miss
+
+TODO
+
+### Out Of Cluster Cache Hit
+
+In the cache hit scenario, for kube-dns we have to use the uncached metrics for negative queries since negative caching is disabled.
+
+TODO
+
+
+So, from a client perspective CoreDNS performs better than kube-dns for out of cluster queries for both cache hits and misses.  Kube-dns performance suffers greatly here because it disables negative caching.
```

Comment:
Note the if this bug was fixed in dnsmasq it wouldn't suffer so badly, so we should be a bit careful in trumping this as a win for CoreDNS (granted right now it *is* a win)

Reply:
Yeah - if we compare with neg-cache enabled, CoreDNS will be significantly worse in all categories.

Reply:
Actually, since I am re-doing all the tests, do think I should test with neg-cache enabled in dnsmasq?  I didn't think this was a "bug" in dnsmasq.  I thought it was just the fact that dnsmasq doesn't have a max neg-cache setting that can override high TTLs from upstream negative responses like coredns does. But if it's actually a bug thats likely to get fixed, then perhaps we should test with it enabled.